<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import plotly.express as px

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''select count(id) -- считаем количество id
            from vacancies --таблица с вакансиями
'''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''select count(id) --считаем количество работадателей по количеству id
            from employers --таблица работодателей
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''select count(id) --считаем количество регионов по количеству id
            from areas --таблица регионов
'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''select count(id) --считаем количество сфер деятельности по количеству id
            from industries --таблица сфер деятельности
'''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

# выводы по предварительному анализу данных
В базе 49197 вакансий, 23501 работодателя, 1362 региона, 294 количества сфер деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''select 
            count(v.id), --посчитаем количество вакансий
            a.name
            from vacancies v --таблица вакансий
                join areas a on v.area_id = a.id --соединяем таблицу вакансий и регионов
            group by a.name -- группируем по названию региона, чтобы узнать количество вакансий в каждом регионе
            order by count(v.id) DESC --сортировка по количеству вакансий по убыванию
'''

In [13]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


count             name
0     5333           Москва
1     2851  Санкт-Петербург
2     2112            Минск
3     2006      Новосибирск
4     1892           Алматы
..     ...              ...
764      1       Тарко-Сале
765      1    Новоаннинский
766      1            Бирск
767      1           Сасово
768      1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''select 
            count(id) -- считаем количество вакансий
            from vacancies 
            where not salary_from is null or not salary_to is null -- добавляем условие, что хотя бы одно поле с зарплатой заполнено 
'''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''select 
            avg(salary_from) salary_from_mean, --среднее значение нижней границы заработной платы
            avg(salary_to) salary_to_mean      --среднее значение верхней границы заработной платы
            from vacancies
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


salary_from_mean  salary_to_mean
0      71064.657901   110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''select 
            schedule,  --тип графика
            employment,--тип трудоустройства
            count(id)  --количество вакансий
            from vacancies
            group by schedule, employment --группировка, чтобы посчитать количество вакансий по типу трудоустройства и графика
            order by count(id) DESC       --сортировка по убыванию количества вакансий
'''

In [19]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''select 
            experience, --опыт работы
            count(id)   --количество вакансий
            from vacancies
            group by experience --группировка по опыту работы, чтобы узнать сколько вакансий ему соответствует
            order by count(id)  --сортировка по возрастанию количества вакансий
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

In [22]:
# выводы по детальному анализу вакансий

Больше всего вакансий в регионах - Москва, Санкт-Петербург, Минск, Новосибирск и Алматы. При обработке датасета можно заменить название регионов, где мало вакансий на одно значение - "Другие".

Больше половины соискателей не указали ни верхней ни нижней границы зарплаты. Можно попробовать заполнить пустые поля медианным значением.

Зарплатная вилка находится в диапазоне от 71тыс до 111тыс.

Больше всего вакансий с типом трудоустройства - полная занятость и графиками работы - полная занятость, удаленная работа, гибкий график.

Больше всего вакансий, где указан опыт работы от 1 года до 3 лет. Менее всего с опытом - более 6 лет, что видимо связано, сто таких специалистов трудно найти.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [23]:
# текст запроса
query_5_1 = f'''
            select 
                e.name,     --наименование работодателя
                count(v.id) --количество вакансий
            from employers e
                join vacancies v on v.employer_id = e.id --соединение с таблицей вакансий
            group by e.name --группировка по работодателю, чтобы узнать количество вакансий, которое он предлагает
            order by count(v.id) desc --сортировка по количеству вакансий
'''

In [24]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection) 
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\690659761.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0                      Яндекс   1933
1                  Ростелеком    491
2                    Тинькофф    444
3                        СБЕР    428
4               Газпром нефть    331
...                       ...    ...
14761            им. Тьюринга      1
14762        LLC Novalab Tech      1
14763   Компания Мир упаковки      1
14764  Лифинцев Олег Иванович      1
14765      ТрансЭнергоХолдинг      1

[14766 rows x 2 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [25]:
# текст запроса
query_5_2 = f'''
            select 
                a.name,
                count(e.id) cnt_emp,
                count(v.id) cnt_vac
            from areas a
                left join vacancies v on v.area_id = a.id --соединяем с таблицей вакансии, левое соединение, чтобы все регионы попали в выборку
                left join employers e on e.area = a.id    --соединяем с таблицей работодателей, левое соединение, чтобы все регионы попали в выборку
            where v.id is null --оставляем регионы, где нет вакансий
            group by a.name       --группиуем по региону, чтобы получить количество работодателей и вакансий в регионе
            order by cnt_emp desc --сортировка по количеству работодателей
'''

In [26]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df 

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\1661138368.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  cnt_emp  cnt_vac
0                            Россия      410        0
1                         Казахстан      207        0
2                Московская область       75        0
3                Краснодарский край       19        0
4                          Беларусь       18        0
..                              ...      ...      ...
588                        Кувандык        0        0
589                        Куйбышев        0        0
590  Куйбышево (Ростовская область)        0        0
591                        Култаево        0        0
592                        Кульсары        0        0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [27]:
# текст запроса
query_5_3 = f'''
            select 
                e.name,
                count(DISTINCT v.area_id) cnt_area --получаем количество регионов, каждый регион считаем один раз
            from vacancies v
                join employers e on e.id = v.employer_id --соединение с таблицей работодателей
            group by e.name        --группировка по работодателю
            order by cnt_area desc --сортируем по количеству
'''

In [28]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df 

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\3759115417.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  cnt_area
0                      Яндекс       181
1                  Ростелеком       152
2                  Спецремонт       116
3      Поляков Денис Иванович        88
4                   ООО ЕФИН         71
...                       ...       ...
14761                  UniSol         1
14762            UNISTORY LLC         1
14763                   UNIT6         1
14764     United Distribution         1
14765              UNITEL LLC         1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [29]:
# текст запроса
query_5_4 = f'''
            select 
                count(e.id) --считаем количество работодателей
            from employers e
                left join employers_industries ei on ei.employer_id = e.id --соединение с таблицей, связывающей сферы деятельности и работодателей
            where ei.employer_id is NULL --оставляем работодателей, у которых не указана сфера деятельности
'''

In [30]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [31]:
# текст запроса
query_5_5 = f'''
            select 
                e.name,
                count(ei.industry_id) cnt_i
            from employers e
                left join employers_industries ei on ei.employer_id = e.id
            group by e.name                  --группируем по названию компании
            having count(ei.industry_id) = 4 --условие для группировки
            order by e.name                  --сортировка по алфавиту
'''

In [32]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  cnt_i
0        101 Интернет      4
1            21vek.by      4
2                2ГИС      4
3                  2К      4
4         4 пикселя +      4
...               ...    ...
1131            ЮРИОН      4
1132  ЮТИП Технологии      4
1133           ЯКласс      4
1134            ЯрНео      4
1135    Яшин&Партнёры      4

[1136 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [33]:
# текст запроса
query_5_6 = f'''
            select 
                count(ei.employer_id) cnt_e,
                i.name
            from employers_industries ei
                join industries i on ei.industry_id = i.id          --соединеняем с таблицей по сферам деятельности, чтобы узнать название сферы
            where i.name like 'Разработка программного обеспечения' --условие на отбор по сфере деятельности
            group by i.name
'''

In [34]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


cnt_e                                 name
0   3553  Разработка программного обеспечения

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [35]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup

#получаем данные со страницы википедии
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text,'html.parser')

#находим нужную таблицу в разметке страницы
cities_here = str(page.find('table', class_="standard")) 
cities_here = cities_here.split('title="')[3:] #получаем список говородов с другой информацией, от который далее избавимся

cities = []
for i in cities_here:
    cities.append(i.split('">')[0]) #получаем только первый элемент, после разделения строки, это и будет нужный нам город
cities = tuple(cities) #преобразуем список в кортеж   

In [36]:
# текст запроса
query_5_7 = f'''
            with temp as( 
            --вложенный запрос, с которым далее работаем, как с таблицей temp
                select 
                    a.name,
                    count(v.id) cnt_v
                from vacancies v
                    join areas a on v.area_id = a.id              --соединение с таблицей регионов, чтобы узнать название региона
                    join employers e on v.employer_id = e.id      --соединение с таблицей работодателей, чтобы узнать название работодателя
                where e.name like 'Яндекс' and a.name in {cities} --работаем только с организацией Яндекс и регионами-городами-милионниками
                group by a.name
            )
            select * from temp --выбираем все из вложенного запроса
            
            UNION 
            
            select
                'Total',
                sum(cnt_v) --считаем общее количество вакансий без группировки по региону
            from temp
            order by 2 desc --сортируем по количеству вакансий
'''

In [37]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt_v
0             Total  485.0
1            Москва   54.0
2   Санкт-Петербург   42.0
3      Екатеринбург   39.0
4   Нижний Новгород   36.0
5       Новосибирск   35.0
6           Воронеж   32.0
7         Краснодар   30.0
8            Самара   26.0
9               Уфа   26.0
10            Пермь   25.0
11           Казань   25.0
12   Ростов-на-Дону   25.0
13        Волгоград   24.0
14       Красноярск   23.0
15        Челябинск   22.0
16             Омск   21.0

***

In [38]:
# выводы по анализу работодателей

Больше всего вакансий предлагает Яндекс, видимо из-за большого количества филиалов в разных городах. Далее по полученному списку тоже идут сети, что может также говорить о текучке кадров в этих компаниях. При обработке данных можно заменить название работодателей с малым количеством вакансий на значение "Остальное", возможно это улучшит модель.

В вакансиях указывают более детальную информацию о регионе, поэтому, например, в регионе "Россия" нет вакансий, но работодатели не указывают в информации о себе подробную информаю о местонахождении. Также нет работодателей и вакансий в маленьких населенных пунктах.

Больше всего разных регионов в вакансиях Яндекса и Ростелеком, их филиалы есть во многих городах и видимо большая текучка кадров, либо они расширяются.

Из 23501 работодателя не указали сферу деятельности 8419, возможно в таблице сфер деятельности нет нужного значения. Возможно стоит заменить пустые поля на значение "Другое", но сначала нужно посмотреть что вообще есть в таблице сфер деятельности.

Количетво работодателей с 4-мя сферами дейтельности 1135.

3553 работодателей из 23501 занимаются разработкой программного обеспечения.

В более крупных городах Яндекс публикует больше вакансий - в Москве и Санкт-Петербурге больше всего вакансий.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [39]:
# текст запроса
query_6_1 = f'''
        select 
            count(id)
        from vacancies
        where lower(name) like '%data%' or lower(name) like '%данн%' --условие, чтобы получить вакансии, где в названии есть данн или data в любои регистре
'''

In [40]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [41]:
# текст запроса
query_6_2 = f'''
            select 
                count(id)
            from vacancies
            --ниже добавляем условия для отбора вакансий для дата-сайентистов
            where (lower(name) like '%data scientist%' 
                or lower(name) like '%data science%'
                or lower(name) like '%исследователь данных%'
                or (upper(name) like '%ML%' and not upper(name) like '%HTML%')
                or lower(name) like '%machine learning%'
                or lower(name) like '%машинн%обучен%')
                and (lower(name) like '%junior%' or experience = 'Нет опыта' or employment = 'Стажировка') --отбираем вакансии уровня джуниор
'''

In [42]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [43]:
# текст запроса
query_6_3 = f'''
            select 
                count(id)
            from vacancies
            --ниже добавляем условия для отбора вакансий для дата-сайентистов
            where (lower(name) like '%data scientist%' 
                or lower(name) like '%data science%'
                or lower(name) like '%исследователь данных%'
                or (name like '%ML%' and not name like '%HTML%')
                or lower(name) like '%machine learning%'
                or lower(name) like '%машинн%обучен%')
                and (upper(key_skills) like '%SQL%' or lower(key_skills) like '%postgres%') --находим те, где указан навык SQL или postgres
'''

In [44]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [45]:
# текст запроса
query_6_4 = f'''
            select 
                count(id)
            from vacancies
            --ниже добавляем условия для отбора вакансий для дата-сайентистов
            where (lower(name) like '%data scientist%' 
                or lower(name) like '%data science%'
                or lower(name) like '%исследователь данных%'
                or (name like '%ML%' and not name like '%HTML%')
                or lower(name) like '%machine learning%'
                or lower(name) like '%машинн%обучен%')
                and (lower(key_skills) like '%python%') --находим те, где указан навык Python
'''

In [46]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [47]:
# текст запроса
query_6_5 = f'''
            select 
                round(avg(length(key_skills) - length(replace(key_skills, CHR(9), ''))+1), 2) as skills_avg
            from vacancies
            --ниже добавляем условия для отбора вакансий для дата-сайентистов
            where (lower(name) like '%data scientist%' 
                or lower(name) like '%data science%'
                or lower(name) like '%исследователь данных%'
                or (name like '%ML%' and not name like '%HTML%')
                or lower(name) like '%machine learning%'
                or lower(name) like '%машинн%обучен%')
                and key_skills is not null --берем только те данные, где указаны навыки 
'''

In [48]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


skills_avg
0        6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [49]:
# текст запроса
query_6_6 = f'''
            select 
                experience,  
                round(avg(coalesce((salary_from + salary_to)/2, salary_from, salary_to)), 0) salary_avg
            from vacancies
            --ниже добавляем условия для отбора вакансий для дата-сайентистов
            where (lower(name) like '%data scientist%' 
                or lower(name) like '%data science%'
                or lower(name) like '%исследователь данных%'
                or (name like '%ML%' and not name like '%HTML%')
                or lower(name) like '%machine learning%'
                or lower(name) like '%машинн%обучен%')
                and (not salary_from is null or not salary_to is null) --выбираем те записи, где указана заработная плата
            group by experience --группируем по опыту   
'''

In [50]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  salary_avg
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

In [51]:
# выводы по предметному анализу

Всего 1771 вакансий имеет отношение к данным. 51 вакансия имеет отношение к дата-сайенсу и подходит для junior-специалистов.
201 вакансия имеет отношение к дата-сайенсу и требует навыков SQL или postgres
351 вакансия имеет отношение к дата-сайенсу и требует знания Python

6,41 в среднем ключевых навыков указывают в вакансиях для DS

Для опыта работы от 3-х до 6-ти лет средняя заработная плата самая большая и составляет 243 тыс, для опыта работы от 1 года до 3-х лет средняя заработная плата составляет 139,5 тыс, самая маленькая средняя заработная плата для специалистов без опыта и составляет 74,5 тыс.

# Общий вывод по проекту

In [52]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования


#посмотрим есть ли вакансии для руководителей
#будем считать, что это для руководителя, если в названии есть "руководит" или "начальник"
query_7_1 = f'''
            select 
                name
            from vacancies
            --ниже добавляем условия для отбора вакансий руководителей
            where (lower(name) like '%руководит%' 
                or lower(name) like '%начальник%')            
'''
# результат запроса
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\2058405612.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


name
0          Начальник отдела информационной безопасности
1     Руководитель внедрения IT-проектов (Дальний Во...
2                           Руководитель отдела дизайна
3           Руководитель группы поддержки пользователей
4                              Руководитель ИТ-проектов
...                                                 ...
2591  Руководитель группы контроля качества (Тестиро...
2592                          Руководитель проекта ТРИЗ
2593         Начальник отдела информационных технологий
2594                     Руководитель процессного офиса
2595                         Начальник сервисной службы

[2596 rows x 1 columns]

In [53]:
#посмотрим среднюю зарплату руководителей в зависимости от опыта
# текст запроса
query_7_2 = f'''
            select 
                experience,  
                round(avg(coalesce((salary_from + salary_to)/2, salary_from, salary_to)), 0) salary_avg
            from vacancies
            --ниже добавляем условия для отбора вакансий руководителей
            where (lower(name) like '%руководит%' 
                or lower(name) like '%начальник%')  
            and (not salary_from is null or not salary_to is null) --выбираем те записи, где указана заработная плата    
            group by experience --группируем по опыту   
'''
# результат запроса
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\659827052.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


experience  salary_avg
0         Более 6 лет    134751.0
1           Нет опыта     62894.0
2  От 1 года до 3 лет     90405.0
3       От 3 до 6 лет    126419.0

2596 из 49197 вакансий относятся к руководителям. Разница между зарплатой в вакансиях с опытом от 3 до 6 лет и более 6 лет небольшая.
Есть вакансии, где нужен руководитель без опыта работы. У руководителей средняя заработная плата ниже средней заработной платы дата-сайентистов.

In [54]:
#узнаем минимальные и максимальные значения заработной платы
# текст запроса
query_7_3 = f'''select 
            min(salary_from) salary_from_min, --минимальное значение нижней границы заработной платы
            max(salary_to) salary_to_max      --максимальное значение верхней границы заработной платы
            from vacancies
'''
# результат запроса
df = pd.read_sql_query(query_7_3, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\1263357928.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_3, connection)


salary_from_min  salary_to_max
0               30        1000000

В базе есть заработные платы с суммой от 30 рублей. Это или ошибка, либо указана оплата за маленький промежуток времени или за единицу услуги, например оплата 30 рублей за одну страницу напечатанного текста, нужно проверить такие значения и заменить их.
Также в базе есть заработная плата в 1млн, что скорее всего является ошибкой.

In [55]:
#узнаем какие сферы деятельности есть в вакансиях
# текст запроса
query_7_4 = f'''
            with temp as( 
            --вложенный запрос, с которым далее работаем, как с таблицей temp
            select 
                count(DISTINCT v.id) cnt_v, --количество вакансий по сфере деятельности
                ei.industry_id
            from vacancies v    
                join employers_industries ei on ei.employer_id = v.employer_id --соединеняем с таблицей по сферам деятельности
            group by ei.industry_id   
            )
            select 
                t.cnt_v,
                i.name 
            from temp t
                join industries i on t.industry_id = i.id --соединеняем с таблицей по сферам деятельности, чтобы узнать название сферы  
            order by 1 DESC
'''
# результат запроса
df = pd.read_sql_query(query_7_4, connection)
df

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\3688696725.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_4, connection)


cnt_v                                               name
0    12499                Разработка программного обеспечения
1    11034  Системная интеграция,  автоматизации технологи...
2     6413  Интернет-компания (поисковики, платежные систе...
3     2742                                               Банк
4     1585                                    Мобильная связь
..     ...                                                ...
288      5                               Спортивная федерация
289      4                                          Химчистка
290      2             Ассоциация в сфере культуры, искусства
291      1                                          Автошкола
292      1                                  Ритуальные услуги

[293 rows x 2 columns]

Больше всего вакансий по разработке программного обеспечения, интеграции и автоматизации.

In [56]:
#посмотрим распределение зарплат по городам-милионникам
# текст запроса
query_7_5 = f'''select 
            (v.salary_from + v.salary_to)/2 salary_avg, --средняя заработная плата
            a.name area
            from vacancies v
                join areas a on a.id = v.area_id
            where a.name in {cities} --отбор по городам   
'''

data_salary = pd.read_sql_query(query_7_5, connection) #отбираем данные с ЗП и регионом 
#строим график
fig = px.box(
    data_frame=data_salary,
    x='area',
    y='salary_avg',
    color='area',
    title='Распределение ЗП по городам-милионникам'
)
fig.show()
fig.write_html("data/image1.html")

C:\Users\tatia\AppData\Local\Temp\ipykernel_21508\475373008.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_salary = pd.read_sql_query(query_7_5, connection) #отбираем данные с ЗП и регионом


Самые высокие зарплаты в Москве, самый маленький разброс зарплат в Челябинске и Красноярске.

Общие выводы:

В базе 49197 вакансий, 23501 работодателя, 1362 региона, 294 количества сфер деятельности.

Больше всего вакансий в крупных городах - Москва, Санкт-Петербург, Минск, Новосибирск и Алматы. При обработке датасета можно заменить название регионов, где мало вакансий на одно значение - "Другие".

Больше половины соискателей не указали ни верхней ни нижней границы зарплаты. Можно попробовать заполнить пустые поля медианным значением.

Больше всего вакансий предлагает Яндекс, видимо из-за большого количества филиалов в разных городах. Далее по полученному списку тоже идут сети, что может также говорить о текучке кадров в этих компаниях. При обработке данных можно заменить название работодателей с малым количеством вакансий на значение "Остальное", возможно это улучшит модель.

Больше всего вакансий по разработке программного обеспечения, интеграции и автоматизации. Также есть сферы деятельности, по которым мало вакансий, возможно стоит заменить эти значения на "Другое".

Далее можно посмотреть зависимость зарплат от опыта работы и графика работы.